include/deal.II/lac/block_linear_operator.h

```diff
@@ -662,6 +662,204 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
   return result * inverse_diagonal;
 }
 
+/**
+ * @relates LinearOperator
+ *
+ * This function implement a forward substitution argument to invert a lower
+ * block triangular matrix.
+ * It takes as argement a block matrix @p block_matrix and an array of LinearOperators
+ * @p inverse_diagonal representing inverses of digonal blocks of @p block_matrix.
+ *
+ * Let us assume we have a linear system where each coefficient of the system is a
+ * matrix:
+ * A00 x0 + ...                   = y0
+ * A01 x0 + A11 x1 + ...          = y1
+ * ...        ...
+ * A0n x0 + A1n x1 + ... + Ann xn = yn
+ *
+ * First fo all, x0 = A00^-1 y0
+ * Then, we can use x0 to recover x1:
+ *    x1 = A11^-1 ( y1 - A01 x0 )
+ * and therefore:
+ *    xn = Ann^-1 ( yn - A0n x0 - ... - A(n-1)n x(n-1) )
+ *
+ * Caveat: Tvmult and Tvmult_add have not been implemented, yet. This may lead to mistakes.
+ * @ingroup LAOperators
+ */
+
+template <unsigned int n, typename Range, typename Domain, typename BlockMatrix>
+LinearOperator<Range, Domain>
+block_forward_substitution(const BlockMatrix &block_matrix,
+                           const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, n> &inverse_diagonal)
```

Comment:
In a call to this declaration, only the template arguments `BlockMatrix` and `n` can be deduced by the compiler, whereas `Range` and `Domain` need to be specified by hand. It may be useful to have these two as the first two template parameters so that only they need to be specified and the trailing template parameters can be deduced automatically.

I imagine `BlockMatrix` can be a linear operator itself, or really anything. Can we ask them for their own `RangeType` and `DomainType`? If so, we could get rid of the `Range` and `Domain` template arguments.